basic_pipeline/3_Source.md

```diff
@@ -0,0 +1,137 @@
+```Elixir
+# FILE: lib/elements/Source.ex
+
+defmodule Basic.Elements.Source do
+ use Membrane.Source
+ alias Membrane.Buffer
+ alias Basic.Formats.Packet
+ ...
+end
+
+```
+# Pads and options
+Later on, we will make use of macros defined in the `Membrane.Source` module:
+```Elixir
+# FILE: lib/elements/Source.ex
+
+defmodule Basic.Elements.Source do
+ ...
+ def_options location: [type: :string, description: "Path to the file"]
+ def_output_pad :output, [caps: {Packet, type: :custom_packets}, mode: :pull]
+ ...
+end
+
+```
+
+The first macro, `def_options` allows us to define the parameters which are expected to be passed while instantiating the element. The parameters will be passed as an automatically generated structure `%Basic.Elements.Source{}`. In our case, we will have a `:location` field inside of that structure. This parameter is about to be a path to the files which will contain input packets.
+Later on, while instatiating the Source element, we will be able to write:
+```Elixir
+%Basic.Elements.Source{location: "input.A.txt"}
+```
+and the `:location` option will be passed during the construction of the element.
+
+The second macro, `def_output_pad`, helps us define the output pad. The pad name will be `:output` (which is a default name for the output pad). The second argument of the macro describes the `caps:` - which is the type of data sent through the pad. As the code states, we want to send data in `Basic.Formats.Packet` format.
+What's more, we have specified that the `:output` pad will work in the `:pull` mode. 
+You can read more on the pad specification [here](https://hexdocs.pm/membrane_core/Membrane.Pad.html#t:common_spec_options_t/0).
+# Initialization of the element
+Let's define our first callback! Why not start with [`handle_init/1`](https://hexdocs.pm/membrane_core/Membrane.Element.Base.html#c:handle_init/1), which gets called once the element is created?
+```Elixir
+# FILE: lib/elements/Source.ex
+
+defmodule Basic.Elements.Source do
+ ...
+ @impl true
+ def handle_init(options) do
+  {:ok,
+    %{
+      location: options.location,
+      content: nil
+    }
+  }
+ end
+ ...
+end
+```
+
+As said before, `handle_init/1` expects a structure with the previously defined parameters to be passed as an argument.
+All we need to do there is to initialize the state - our state will be in a form of a map, and for now on we will put there a `location` (a path to the input file) and the `content`, where we will be holding packets read from the file, which haven't been sent yet. For now, the content is set to nil as we haven't read anything from the input file yet.
+You might also wonder what is the purpose of the `@impl true` specifier, put just above the function signature - this is simply a way to tell the compiler that the function defined below is about to be a callback. If we have misspelled the function name (or provided a wrong arguments list), we will be informed in the compilation time.
```

Comment:
We can remove that, or mark it as some tip section.

Reply:
done